internal/oidc/login/post_login_handler.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2022 the Pinniped contributors. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package login
+
+import (
+	"net/http"
+	"net/url"
+
+	"github.com/ory/fosite"
+
+	"go.pinniped.dev/internal/httputil/httperr"
+	"go.pinniped.dev/internal/oidc"
+	"go.pinniped.dev/internal/oidc/downstreamsession"
+	"go.pinniped.dev/internal/plog"
+)
+
+func NewPostHandler(issuerURL string, upstreamIDPs oidc.UpstreamIdentityProvidersLister, oauthHelper fosite.OAuth2Provider) HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request, encodedState string, decodedState *oidc.UpstreamStateParamData) error {
+		// Note that the login handler prevents this handler from being called with OIDC upstreams.
+		_, ldapUpstream, idpType, err := oidc.FindUpstreamIDPByNameAndType(upstreamIDPs, decodedState.UpstreamName, decodedState.UpstreamType)
+		if err != nil {
+			// This shouldn't normally happen because the authorization endpoint ensured that this provider existed
+			// at that time. It would be possible in the unlikely event that the provider was deleted during the login.
+			plog.Error("error finding upstream provider", err)
+			return httperr.Wrap(http.StatusUnprocessableEntity, "error finding upstream provider", err)
+		}
+
+		// Get the original params that were used at the authorization endpoint.
+		downstreamAuthParams, err := url.ParseQuery(decodedState.AuthParams)
+		if err != nil {
+			// This shouldn't really happen because the authorization endpoint encoded these query params correctly.
```

Comment:
> This was a choice that we (you and I) made

We all make mistakes :)

> Doesn't seem like there's any particular reason to change it at this point

Parsing the same string multiple times per request is hardly ideal, and we are changing the format anyway so it would not hurt to fix it here.  But not a big deal either.